stratosphere/creport/source/creport_threads.cpp

```diff
@@ -0,0 +1,267 @@
+            }
+        }
+        if (this->tls_address != 0) {
+            fprintf(f_report, "    TLS Address:                 %016lx\n", this->tls_address);
+            fprintf(f_report, "    TLS Dump:                                 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f\n");
+            for (size_t i = 0; i < 0x10; i++) {
+                const size_t ofs = i * 0x10;
+                fprintf(f_report, "                                 %012lx %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x\n",
+                    this->tls_address + ofs, this->tls[ofs + 0], this->tls[ofs + 1], this->tls[ofs + 2], this->tls[ofs + 3], this->tls[ofs + 4], this->tls[ofs + 5], this->tls[ofs + 6], this->tls[ofs + 7],
+                    this->tls[ofs + 8], this->tls[ofs + 9], this->tls[ofs + 10], this->tls[ofs + 11], this->tls[ofs + 12], this->tls[ofs + 13], this->tls[ofs + 14], this->tls[ofs + 15]);
+            }
+        }
+    }
+
+    bool ThreadInfo::ReadFromProcess(Handle debug_handle, std::map<u64, u64> &tls_map, u64 thread_id, bool is_64_bit) {
+        /* Set thread id. */
+        this->thread_id = thread_id;
+
+        /* Verify that the thread is running or waiting. */
+        {
+            u64 _;
+            u32 _thread_state;
+            if (R_FAILED(svcGetDebugThreadParam(&_, &_thread_state, debug_handle, this->thread_id, DebugThreadParam_State))) {
+                return false;
+            }
+
+            const svc::ThreadState thread_state = static_cast<svc::ThreadState>(_thread_state);
+            if (thread_state != svc::ThreadState::Waiting && thread_state != svc::ThreadState::Running) {
+                return false;
+            }
+        }
+
+        /* Get the thread context. */
+        if (R_FAILED(svcGetDebugThreadContext(&this->context, debug_handle, this->thread_id, svc::ThreadContextFlag_All))) {
+            return false;
+        }
+
+        /* In aarch32 mode svcGetDebugThreadContext does not set the LR, FP, and SP registers correctly. */
+        if (!is_64_bit) {
+            this->context.fp = this->context.cpu_gprs[11].x;
+            this->context.sp = this->context.cpu_gprs[13].x;
+            this->context.lr = this->context.cpu_gprs[14].x;
+        }
+
+        /* Read TLS, if present. */
+        /* TODO: struct definitions for nnSdk's ThreadType/TLS Layout? */
+        if (tls_map.find(thread_id) != tls_map.end()) {
+            this->tls_address = tls_map[thread_id];
+            u8 thread_tls[0x200];
+            if (R_SUCCEEDED(svcReadDebugProcessMemory(thread_tls, debug_handle, this->tls_address, sizeof(thread_tls)))) {
+                std::memcpy(this->tls, thread_tls, sizeof(this->tls));
+                /* Try to detect libnx threads, and skip name parsing then. */
+                if (*(reinterpret_cast<u32 *>(&thread_tls[0x1E0])) != LibnxThreadVarMagic) {
```

Comment:
These casts are ugly. Would it be worth making a templated util function to read an arbitrary value at an arbitrary address? Something like:
```cpp
namespace {
    template <typename T>
    T Read(const void* ptr) {
        return *(reinterpret_cast<const T *>(ptr));
    }
}
```
and it would be used like this: `Read<u32>(&thread_tls[0x1E0])`

Or alternatively, wrap tls accessors/parsers in utility functions so that they're not in the main logic.